colmeia-hypercore/src/network.rs

```diff
@@ -0,0 +1,294 @@
+        + std::fmt::Debug
+        + Send
+        + Sync
+        + 'static,
+{
+    let (metadata_sender, mut metadata_receiver) = futures::channel::mpsc::unbounded();
+    let mut metadata_job = Some(JobHolder::Sender(metadata_sender));
+
+    let (content_sender, mut content_receiver) = futures::channel::mpsc::unbounded();
+    let mut content_job = Some(JobHolder::Sender(content_sender));
+
+    loop {
+        let (event, _, _) = futures::future::select_all(vec![
+            client.loop_next().map(HyperdriveEvents::Client).boxed(),
+            metadata_receiver
+                .select_next_some()
+                .map(HyperdriveEvents::Metadata)
+                .boxed(),
+            content_receiver
+                .select_next_some()
+                .map(HyperdriveEvents::Content)
+                .boxed(),
+        ])
+        .await;
+
+        match event {
+            HyperdriveEvents::Client(Ok(proto::Event::DiscoveryKey(message))) => {
+                if let Some(JobHolder::Sender(_)) = &metadata_job {
+                    let drive = hyperdrive.read().await;
+                    let metadata = drive.metadata.read().await;
+                    let public_key_for_metadata = metadata.public_key().as_bytes();
+                    let metadata_discovery_key =
+                        hypercore_protocol::discovery_key(public_key_for_metadata);
+                    if message == metadata_discovery_key.as_slice()
+                        && client.open(public_key_for_metadata.to_vec()).await.is_err()
+                    {
+                        log::error!("failed to open metadata data exchange");
+                        break;
+                    };
+                }
+            }
+            HyperdriveEvents::Client(Ok(proto::Event::Channel(channel))) => {
+                // TODO don't rely on channel event oerder - check the keys to see if they match
+                if let Some(JobHolder::Sender(sender)) = metadata_job.take() {
+                    log::debug!("initializing metadata feed");
+                    let feed = hyperdrive.read().await.metadata.clone();
+                    metadata_job = Some(JobHolder::Job(task::spawn(replicate_channel(
+                        channel, feed, sender,
+                    ))));
+                    continue;
+                }
+                if let Some(JobHolder::Sender(sender)) = content_job.take() {
+                    log::debug!("initializing content feed");
+                    let feed = hyperdrive.read().await.content.clone();
+                    if let Some(feed) = feed {
+                        content_job = Some(JobHolder::Job(task::spawn(replicate_channel(
+                            channel, feed, sender,
+                        ))));
+                    }
+                    continue;
+                }
+            }
+            HyperdriveEvents::Client(Err(_)) => {
+                // TODO delete this branch
```

Comment:
TODO: Delete this. This is responsability for outside this lib. Caller is the one interested with replication finishing, this is just for debugging.